File: app/src/debug/java/com/example/DebugApp.java
package com.example;

import android.os.StrictMode;

import com.example.android.trackers.ActivityLifecycleTracker;
import com.example.android.trackers.FragmentLifecycleTracker;
import com.example.android.trackers.FrescoCacheStatsTracker;
import com.example.logger.CLog;
import com.facebook.imagepipeline.core.ImagePipelineConfig;
import com.facebook.stetho.Stetho;
import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;
import com.tspoon.traceur.Traceur;

public class DebugApp extends App {
	@Override
	public void onCreate() {
		if (LeakCanary.isInAnalyzerProcess(this)) {
			return;
		}

		super.onCreate();

		Stetho.initializeWithDefaults(this);
		trackActivitiesAndFragmentsLifecycle();
		Traceur.enableLogging();

		enableStrictMode();
	}

	@Override
	public void onTrimMemory(int level) {
		super.onTrimMemory(level);
		CLog.d("[ onTrimMemory ] " + level);
	}

	@Override
	protected RefWatcher enableLeakCanary() {
		return LeakCanary.refWatcher(this).buildAndInstall();
	}

	private void enableStrictMode() {
		StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
			.detectAll()
			.penaltyFlashScreen()
			.penaltyLog()
			.penaltyDeath()
			.build());

		StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
			.detectAll()
			.penaltyLog()
			.penaltyDeath()
			.build());
	}

	@Override
	protected void trackFresco(ImagePipelineConfig.Builder builder) {
		builder.setImageCacheStatsTracker(new FrescoCacheStatsTracker());
	}

	private void trackActivitiesAndFragmentsLifecycle() {
		final FragmentLifecycleTracker fragmentLifecycleTracker = new FragmentLifecycleTracker();
		final ActivityLifecycleTracker activityLifecycleTracker = new ActivityLifecycleTracker(fragmentLifecycleTracker);

		registerActivityLifecycleCallbacks(activityLifecycleTracker);
	}
}

File: app/src/debug/kotlin/br/com/vitorsalgado/example/DebugApp.kt
package br.com.vitorsalgado.example

import android.os.StrictMode
import br.com.vitorsalgado.example.trackers.ActivityDebugLifecycleTracker
import br.com.vitorsalgado.example.trackers.FragmentDebugLifecycleTracker
import br.com.vitorsalgado.example.utils.LogUtility
import com.facebook.stetho.Stetho
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import com.tspoon.traceur.Traceur
import timber.log.Timber

class DebugApp : App() {
  override fun onCreate() {
    if (LeakCanary.isInAnalyzerProcess(this)) {
      return
    }

    super.onCreate()

    Stetho.initialize(
      Stetho.newInitializerBuilder(this)
        .enableDumpapp(Stetho.defaultDumperPluginsProvider(this))
        .enableWebKitInspector(Stetho.defaultInspectorModulesProvider(this))
        .build()
    )

    trackActivitiesAndFragmentsLifecycleForDebug()
    Traceur.enableLogging()
    Timber.plant(Timber.DebugTree())

    enableStrictMode()
  }

  override fun onTrimMemory(level: Int) {
    super.onTrimMemory(level)
    LogUtility.d("[ onTrimMemory ] $level")
  }

  override fun enableLeakCanary(): RefWatcher {
    return LeakCanary.refWatcher(this).buildAndInstall()
  }

  private fun enableStrictMode() {
    StrictMode.setThreadPolicy(
      StrictMode.ThreadPolicy.Builder()
        .detectAll()
        .penaltyFlashScreen()
        .penaltyLog()
        .build()
    )

    StrictMode.setVmPolicy(
      StrictMode.VmPolicy.Builder()
        .detectAll()
        .penaltyLog()
        .build()
    )
  }

  private fun trackActivitiesAndFragmentsLifecycleForDebug() {
    val fragmentLifecycleTracker = FragmentDebugLifecycleTracker()
    val activityLifecycleTracker = ActivityDebugLifecycleTracker(fragmentLifecycleTracker)

    registerActivityLifecycleCallbacks(activityLifecycleTracker)
  }
}
